Feature/decimal output

BishopT opened the pull request:
Decimal output was needed on my setup in order to use as arguments for riverctl

tarneaux commented:
I have cherry-picked the first commit, but I'm not really fund of the second one as it introduces a breaking change: if someone is using this script and just taking the second word from every line, they're going to get confused at why it suddenly outputs a number instead of its binary representation.

Do you need both the binary and the decimal representation? If not, an option could be added (something like `-d`) to make the output decimal instead of binary (removing the binary printing). But this means adding a whole argument parser, and since I went back to X11 to let river mature a bit more before using it, I can't test any changes I would be making.

Could you implement an argument parser as described?

tarneaux commented:
An alternative to using a `-d` argument and just not printing the binary representation, a `-f` argument with a string parameter could be added instead. The parameter would contain `b` and/or `d`, in combination with optional spaces, to specify the format of the ouptut, e.g.:

`-f "b d"` => `0011 3`

tarneaux commented:
Adding an argument parser could also allow the user to specify the number of printed bits instead of editing the code, with a `-c` argument.

BishopT commented:
I agree and I had already in mind the command line args as a better long term solution.  
This was more a quick and dirty implementation to test out the setup. I proposed the PR as I was not sure to be able to go further in implementing the args and I admitedly didn't had in mind this would break existing setups.
I think I'll go by steps, implementing only the `-f d` option as a first step.

BishopT commented:
build & test report for commit [b3771bf](https://github.com/tarneaux/river-get-tags/pull/1/commits/b3771bfd9fffc2393a9abc02cf94dcff87c47344):
[build.log](https://github.com/tarneaux/river-get-tags/files/14783260/build.log)
[tests.log](https://github.com/tarneaux/river-get-tags/files/14783259/tests.log)

tarneaux commented:
Thanks! I like the way you did it, but I have a few issues with it:
- Could you make the arg parser a separate function, and use a struct to store the arguments?
- The format could have some default values which are changed when an argument is parsed. This would allow avoiding the use of different logic depending on whether or not arguments are passed to the program, just setting the old behavior to be the default.
- Why did you make `fflag`, `oflag` and `bflag` ints? Shouldn't they just be bools?
- I would prefer erroring out to silently capping the number of printed bits (see code below):
```
blength =
  (iarg < MAX_BITFIELD_LENGTH) ? iarg : MAX_BITFIELD_LENGTH;
```
- Shouldn't the format be just one option for both the occupied and focused tags, instead of setting them separately? Can you think of a case where this would be needed?
- What I had in mind for the format option was to allow setting any combination of formats, for example printing both binary and decimal values separated by a space. This PR would make these formats exclusive as of now.

BishopT commented:
Hey, I'm just reviewing your comment right now.

> Could you make the arg parser a separate function, and use a struct to store the arguments?
> The format could have some default values which are changed when an argument is parsed. This would allow avoiding the use of different logic depending on whether or not arguments are passed to the program, just setting the old behavior to be the default.

I think so. What I can think of:
- create parser function taking as input, args and struct, modifying the struct as per parsed args
- in main, declare the struct with default values then call parser 

> Why did you make fflag, oflag and bflag ints? Shouldn't they just be bools?

I was just mimicking some examples. I think they were designed to managed cases where multiples instances of same flag are detected. 

> I would prefer erroring out to silently capping the number of printed bits (see code below):

You're right, it's bad software design to not notify user of wrong inputs. Also capping might not be satisfying as a fallback value. And what if provided value is negative?   
So I think of making the parser check if bitfield length is [1-32] and if not, keep default value + print error message. 

> Shouldn't the format be just one option for both the occupied and focused tags, instead of setting them separately? Can you think of a case where this would be needed?

Not really, I made it like this because it was easy to give the format as arg to the flag. And also because I had a conflict of flag letter for both "focused tags" and "format" options 😅
If the format have to be common and someone needs different format, one still can make 2 call to river-get-tags with different option set. 


So what do you think of :

- `-f` trigger printing of focused tags ; no arg associated to it
- `-o` trigger printing of occupied tags ; no arg associated to it
- `-n` print tags name (focused or occupied) before the value 
- `-F <format>` specify print format ; <format> being (subset of printf options)
  - `[1-32]b` for bitfield
  - `d|i` for integer
  - `[#][1-8]x|X` for hexadecimal

If no arg is provided, the main function will populate with default args: `-fonF5b`

tarneaux commented:
Nice! Could you please update the README and the help string with a few usage examples?
I just think `C printf format with specifiers compatible with unsigned integers (d, i, u, o, x, X)` requires a bit too much guessing.

BishopT commented:
Done. I've also made a few changes:

- when writing help examples, it seems to me it was useless to support signed decimal integers (and prone to errors if one actually uses 32 bits of the bitfield) so I removed them from the list of authorized specifiers
- I thought the default bitfield length was 5 but when updating README file, I saw it was 6 (the `<=5` in the for-loop makes sense to me now ; I thought it was a bug when I added the length parameter). So I put back 6 as default value.

BishopT commented:
[test.log](https://github.com/tarneaux/river-get-tags/files/14898616/test.log)

tarneaux commented:
It's better when I actually merge the whole PR :-)

Thanks!
